### PR TITLE
RFC - Mailer

### PR DESCRIPTION
**Pull request by jadb**
This supersedes #6319 and as suggested by @markstory, adds some conventions borrowed from Rails' [ActionMailer](http://edgeguides.rubyonrails.org/action_mailer_basics.html).
## TODO
- [x] Move all email related classes out of `Cake\Network\Email` and into their own namespace. Agreed with @markstory and @lorenzo on `Cake\Mailer`, with aliases for BC.
- [x] Move transports into their own namespace at `Cake\Mailer\Transport` with aliases for BC.
- [x] Create a new `Mailer` class that would make working with emails very similar to controllers (actions, templates, model-aware).
- [ ] Create new bake command(s) to create mailer classes and/or templates.
- [x] Update cakephp/app's `config/bootstrap.php` and `config/app.default.php`.
## Usage

For documentation's sake and to better grasp the changes, here's a sample code of what would/should work when the work is finished.

```
bin/cake bake mailer user welcome,reset_password
```

which would generate the following:

```
- src/Mailer/UserMailer.php
- src/Template/Layout/Email/html/user.ctp
- src/Template/Layout/Email/text/user.ctp
- src/Template/Email/html/User/welcome.ctp
- src/Template/Email/html/User//reset_password.ctp
- src/Template/Email/text/User/welcome.ctp
- src/Template/Email/text/User/reset_password.ctp
```

`UserMailer` would look like this after some customizations:

``` php
<?php
namespace App\Mailer;

use ArrayObject;
use Cake\Event\Event;
use Cake\Mailer\Mailer;
use Cake\ORM\Entity;

class UserMailer extends Mailer
{
    public function welcome($user)
    {
        $this->subject = sprintf('Welcome %s', $user->name);
        $this->to = $user->email;
    }

    public function resetPassword($user)
    {
        $this->subject = sprintf('Reset password');
        $this->to = $user->email;
        $this->set(['token' => $user->token]);
    }

    public function implementedEvents()
    {
        return [
            'Model.afterSave' => 'onRegistration',
        ];
    }

   public function onRegistration(Event $event, Entity $entity, ArrayObject $options)
   {
       if ($entity->isNew()) {
            $this->send('welcome', [$entity]);
        }
    }
}
```

And then from the controller (or any class using `MailerAwareTrait`):

``` php
<?php
namespace App\Controller;

use Cake\Mailer\MailerAwareTrait;

class UsersController extends AppController
{
    use MailerAwareTrait;

    public function register()
    {
        ...
        $this->getMailer('User')->send('welcome', [$user]);
        ...
    }
}
```

or in a table using events:

``` php
<?php
namespace App\Model\Table;

use App\Mailer\UserMailer;
use Cake\ORM\Table;

class UsersTable extends Table
{
    public function initialize(array $config = [])
    {
        ...
        $this->eventManager->on(new UserMailer());
    }
}
```

### CODE REVIEW

**Comment by Xety** [comment]
:+1: I like it.

File: config/bootstrap.php [hunk 1]
```diff
@@ -19,6 +19,11 @@
 
 // @deprecated Backward compatibility with 2.x series
 if (PHP_VERSION_ID < 70000) {
+    class_alias('Cake\Mailer\AbstractTransport', 'Cake\Network\Email\AbstractTransport');
+    class_alias('Cake\Mailer\Transport\DebugTransport', 'Cake\Network\Email\DebugTransport');
+    class_alias('Cake\Mailer\Email', 'Cake\Network\Email\Email');
+    class_alias('Cake\Mailer\Transport\MailTransport', 'Cake\Network\Email\MailTransport');
+    class_alias('Cake\Mailer\Transport\SmtpTransport', 'Cake\Network\Email\SmtpTransport');
```
Comment (lorenzo) [inline config/bootstrap.php]: these need to be outside the php version check

File: src/Mailer/Email.php [hunk 2]
```diff
@@ -1023,7 +1023,7 @@ protected function _constructTransport($name)
             return $config['className'];
         }
 
-        $className = App::className($config['className'], 'Network/Email', 'Transport');
+        $className = App::className($config['className'], 'Mailer/Transport', 'Transport');
```
Comment (markstory) [inline src/Mailer/Email.php]: We should check the new and then old namespaces. Otherwise userland code will break.

**Comment by markstory** [comment]
I like where this is going. Did you consider have `send('welcome', $user)` instead of using `__call`?

File: src/Mailer/Email.php (hunk 2 above)
Comment (jadb) [inline src/Mailer/Email.php]: Noted. Will make necessary changes.

File: config/bootstrap.php (hunk 1 above)
Comment (jadb) [inline config/bootstrap.php]: Noted.

**Comment by jadb** [comment]
@markstory no but I like it. I stuck to the `ActionMailer` API which only passes `$headers` to `send()`.

File: src/Mailer/Email.php [hunk 3]
```diff
@@ -262,7 +262,7 @@ class Email implements JsonSerializable, Serializable
     /**
      * The transport instance to use for sending mail.
      *
-     * @var \Cake\Network\Email\AbstractTransport
+     * @var \Cake\Mailer\Transport\AbstractTransport
```
Comment (antograssiot) [inline src/Mailer/Email.php]: \Cake\Mailer\AbstractTransport instead?

File: src/Mailer/Email.php [hunk 4]
```diff
@@ -977,7 +977,7 @@ public function emailFormat($format = null)
      *
      * @param string|AbstractTransport|null $name Either the name of a configured
      *   transport, or a transport instance.
-     * @return \Cake\Network\Email\AbstractTransport|$this
+     * @return \Cake\Mailer\Transport\AbstractTransport|$this
```
Comment (antograssiot) [inline src/Mailer/Email.php]: here also \Cake\Mailer\AbstractTransport

File: src/Mailer/Email.php (hunk 3 above)
Comment (jadb) [inline src/Mailer/Email.php]: Good catch!

File: src/Mailer/Email.php (hunk 2 above)
Comment (lorenzo) [inline src/Mailer/Email.php]: I think it is fine if the break as long a we have some sort of upgrade command that can help you move the classes.

Comment (markstory) [inline src/Mailer/Email.php]: But we don't have to have a break if we include 3 extra lines of code.

**Comment by jeremyharris** [comment]
That is some tasty goodness right there. Tired of rewriting this functionality. Nice job!

**Comment by HavokInspiration** [comment]
Looking at this and the other PR that was the starting point, I can already see myself implementing this in my apps. This is really clever.

File: src/Mailer/Exception/MissingMailException.php [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+ */
+namespace Cake\Mailer\Exception;
+
+use Cake\Core\Exception\Exception;
+
+/**
+ * Missing Action exception - used when a controller action
+ * cannot be found, or when the controller's isAction() method returns false.
+ */
+class MissingMailException extends Exception
```
Comment (markstory) [inline src/Mailer/Exception/MissingMailException.php]: The doc comment doesn't match what the code does.

File: src/Mailer/Mailer.php [hunk 6]
```diff
@@ -0,0 +1,222 @@
...
+     * @param array $headers Headers to set.
+     * @return array
+     * @throws \Cake\Mailer\Exception\MissingMailException
+     * @throws \BadMethodCallException
+     */
+    public function send($mail, $args = [], $headers = [])
+    {
+        if (!method_exists($this, $mail)) {
+            throw new MissingMailException([
+                'mailer' => $this->getName() . 'Mailer',
```
Comment (markstory) [inline src/Mailer/Mailer.php]: We can call the mailer methods actions.

File: tests/TestCase/Mailer/MailerTest.php [hunk 7]
```diff
@@ -0,0 +1,118 @@
+<?php
+namespace Cake\Test\TestCase\Mailer;
```
Comment (markstory) [inline tests/TestCase/Mailer/MailerTest.php]: File header missing.

File: tests/TestCase/Mailer/MailerTest.php [hunk 8]
```diff
@@ -0,0 +1,118 @@
...
+
+use Cake\Mailer\Email;
+use Cake\Mailer\Mailer;
+use Cake\TestSuite\TestCase;
+
+class TestMailer extends Mailer
+{
+    public function getEmailForAssertion()
```
Comment (markstory) [inline tests/TestCase/Mailer/MailerTest.php]: Why not have a public getter for the underlying email?

Comment (jadb) [inline tests/TestCase/Mailer/MailerTest.php]: I am not a fan of giving access to unneeded stuff but we could. Aside from tests, I don't see why one would need it using the mailer. The goal was to isolate mail related stuff for easier testing, etc.

**Comment by makallio85** [comment]
Great job. Waiting for it.

File: tests/TestCase/Mailer/MailerTest.php (hunk 8 above)
Comment (markstory) [inline tests/TestCase/Mailer/MailerTest.php]: Ok, we've had problems with rotting test stubs in the past which is why I personally try to avoid them.

Comment (jadb) [inline tests/TestCase/Mailer/MailerTest.php]: Ok, will look into improving that if possible then.

File: src/Mailer/Mailer.php [hunk 9]
```diff
@@ -0,0 +1,222 @@
...
+     * @return $this object.
+     */
+    public function reset()
+    {
+        $this->_email->reset();
+        return $this;
+    }
+
+    /**
+     * {@inheritdoc}
```
Comment (ADmad) [inline src/Mailer/Mailer.php]: I don't think apigen is going to be able to inherit docblocks for php's api docs :)

File: tests/TestCase/Mailer/MailerTest.php [hunk 10]
```diff
@@ -0,0 +1,128 @@
...
+ * @since         3.1.0
+ * @license       http://www.opensource.org/licenses/mit-license.php MIT License
+ */
+namespace Cake\Test\TestCase\Mailer;
+
+use Cake\Mailer\Email;
+use Cake\Mailer\Mailer;
+use Cake\TestSuite\TestCase;
+
+class TestMailer extends Mailer
```
Comment (ADmad) [inline tests/TestCase/Mailer/MailerTest.php]: Would be better to put this class in test app.

File: src/Mailer/Mailer.php [hunk 11]
```diff
@@ -0,0 +1,222 @@
...
+    }
+
+    /**
+     * Returns the mailer's name.
+     *
+     * @return string
+     */
+    public function getName()
+    {
+        return str_replace('Mailer', '', join('', array_slice(explode('\\', get_class($this)), -1)));
```
Comment (ADmad) [inline src/Mailer/Mailer.php]: If we add a public property `Mailer::$name` name could be set statically and this string manipulation would be required only if property is null.

File: src/Mailer/Mailer.php [hunk 12]
```diff
@@ -0,0 +1,222 @@
...
+    {
+        if ($email === null) {
+            $email = new Email();
+        }
+
+        if ($this->layout === null) {
+            $this->layout = Inflector::underscore($this->getName());
+        }
+
+        $this->_email = $email->profile((array)$this);
```
Comment (ADmad) [inline src/Mailer/Mailer.php]: Since `send()` always does `$this->_email->profile((array)$this)` doing it constructor seems redundant.

**Comment by jippi** [comment]
Super late question

Is `Mailer` the right namespace name?

We use `Controller`, `Model`, `Table`, `Template`,`Shell` etc, singular word for almost all directories, based on that, shouldn't it be just `Mail` ?

**Comment by dereuromark** [comment]
I agree with @jippi Mail sounds better.

**Comment by markstory** [comment]
So `Cake\Mail\Email` and `Cake\Mail\Mailer`?

**Comment by jadb** [comment]
To clear things up a little. The namespace discussion was done over IRC with @markstory and @lorenzo present. Mailers work very similarly to controllers, hence also their methods being referred to as `actions` in exceptions and coming documentation. They are borrowed from Ruby's ActionMailer.

@jippi:
> We use Controller, Model, Table, Template,Shell etc, singular word for almost all directories, based on that, shouldn't it be just Mail ?

Mailer is singular.

> Also Validation vs. Validator - if it should be named Mailer

`Mailer` vs `Mailing`? :stuck_out_tongue: 

@dereuromark:

> Mail/Mail (like View/View vs Viewer) maybe sounds better.

Or `Control/Control` (like View/View) or maybe `View\Help`? :de:

**Comment by dereuromark** [comment]
@jadb You convinced me :)

File: src/Mailer/Email.php (hunk 2 above)
Comment (jadb) [inline src/Mailer/Email.php]: I fixed this but it's not showing as outdated. See lines right below it.

**Comment by jadb** [comment]
Bake command added, just missing the `actions`.

File: config/bootstrap.php [hunk 13]
```diff
@@ -22,6 +22,13 @@
     class_alias('Cake\Utility\Text', 'Cake\Utility\String');
 }
 
+// @deprecated Backward compatibility with 2.x, 3.0.x
+class_alias('Cake\Mailer\AbstractTransport', 'Cake\Network\Email\AbstractTransport');
+class_alias('Cake\Mailer\Transport\DebugTransport', 'Cake\Network\Email\DebugTransport');
+class_alias('Cake\Mailer\Email', 'Cake\Network\Email\Email');
+class_alias('Cake\Mailer\Transport\MailTransport', 'Cake\Network\Email\MailTransport');
+class_alias('Cake\Mailer\Transport\SmtpTransport', 'Cake\Network\Email\SmtpTransport');
```
Comment (lorenzo) [inline config/bootstrap.php]: I wonder if we should add this to the app template instead of just to the migration guide, it is a shame having to load all this classes just because.

File: tests/test_app/TestApp/Mailer/TestMailer.php [hunk 14]
```diff
@@ -0,0 +1,32 @@
...
+ * CakePHP(tm) Tests <http://book.cakephp.org/2.0/en/development/testing.html>
+ * Copyright (c) Cake Software Foundation, Inc. (http://cakefoundation.org)
+ *
+ * Licensed under The MIT License
+ * For full copyright and license information, please see the LICENSE.txt
+ * Redistributions of files must retain the above copyright notice
+ *
+ * @copyright     Copyright (c) Cake Software Foundation, Inc. (http://cakefoundation.org)
+ * @link          http://book.cakephp.org/2.0/en/development/testing.html CakePHP(tm) Tests
+ * @since         1.3.0
```
Comment (lorenzo) [inline tests/test_app/TestApp/Mailer/TestMailer.php]: 3.1.0

File: src/Mailer/Email.php [hunk 15]
```diff
@@ -1023,7 +1023,11 @@ protected function _constructTransport($name)
             return $config['className'];
         }
 
-        $className = App::className($config['className'], 'Network/Email', 'Transport');
+        $className = App::className($config['className'], 'Mailer/Transport', 'Transport');
+        if (!$className) {
+            $className = App::className($config['className'], 'Network/Email', 'Transport');
```
Comment (lorenzo) [inline src/Mailer/Email.php]: maybe add a trigger_error() here?

File: config/bootstrap.php (hunk 13 above)
Comment (markstory) [inline config/bootstrap.php]: Does class_alias() actually load the classes?

Comment (ceeram) [inline config/bootstrap.php]: In the example code it will autoload, a 3rd param of class_alias would be needed to set to false to turn off autoloading

Comment (jadb) [inline config/bootstrap.php]: I personally prefer @lorenzo's suggestion of moving this stuff to the app template, however, worries me for already created apps which would need to manually update or having breaking changes to deal with. 

If we want, as @ceeram suggested, we can pass a 3rd param to control autoloading but why aren't we doing it for the `Text` class as well?

Comment (ceeram) [inline config/bootstrap.php]: Moving this to app-template serves no purpose if we make the app-template require correct cakephp version which then does not need any aliassing.
The aliassing is needed in core for BC to applications using the current classnames.

Comment (ADmad) [inline config/bootstrap.php]: :+1: to @ceeram's suggestion and doing same for `Text` class too.

Comment (markstory) [inline config/bootstrap.php]: Moving the aliasing to the app template doesn't help people when they upgrade. Adding the 3rd as @ceeram suggested sounds good though.

**Comment by markstory** [comment]
Looks like only a few things left to fix up before this can be merged @jadb

**Comment by jadb** [comment]
Yes, will wrap it up this week.

**Comment by jadb** [comment]
Took care of all comments. The only thing I had planned (wishful thinking) was the bake commands but didn't get time to write those. Let me know if they are required for this to be merged or if they could be split into an issue of their own.

**Comment by markstory** [comment]
Bake commands would be a separate pull request anyways as bake is a plugin.

**Comment by jadb** [comment]
I agree (like cakephp/app) but was just a part of what I had listed to be done so didn't want to just drop the ball on it without flagging it I guess.

**Comment by antoniovassell** [comment]
+1

**Comment by ADmad** [comment]
Bake templates can be done later. This is good to merge.

File: src/Mailer/Mailer.php [hunk 16]
```diff
@@ -0,0 +1,249 @@
...
+    /**
+     * Gets the property value if it exists.
+     *
+     * @param string $offset Property name.
+     * @return mixed Value.
+     */
+    public function offsetGet($offset)
+    {
+        if (!$this->offsetExists($offset)) {
+            throw \Exception();
```
Comment (markstory) [inline src/Mailer/Mailer.php]: This exception should have a message.

File: tests/TestCase/Mailer/MailerTest.php [hunk 17]
```diff
@@ -0,0 +1,121 @@
...
+        $email->expects($this->once())
+            ->method('addHeaders')
+            ->with([]);
+        $result = (new TestMailer($email))->addHeaders([]);
+        $this->assertInstanceOf('TestApp\Mailer\TestMailer', $result);
+
+        $email = $this->getMockForEmail('attachments');
+        $email->expects($this->once())
+            ->method('attachments')
+            ->with([]);
```
Comment (markstory) [inline tests/TestCase/Mailer/MailerTest.php]: It might be good to check these with non empty arguments to ensure arguments are passed correctly.

File: tests/test_app/TestApp/Template/Email/html/html.ctp [hunk 18]
```diff
@@ -1,5 +1,5 @@
 <?php
-use Cake\Network\Email\Email;
+use Cake\Mailer\Email;
```
Comment (markstory) [inline tests/test_app/TestApp/Template/Email/html/html.ctp]: Why is this here?

Comment (jadb) [inline tests/test_app/TestApp/Template/Email/html/html.ctp]: Hmm - I guess with find and replace in entire project for namespace, there was a reference to `Cake\Network\Email\Email`.

File: src/Mailer/Mailer.php (hunk 16 above)
Comment (jadb) [inline src/Mailer/Mailer.php]: Will do.

File: tests/TestCase/Mailer/MailerTest.php (hunk 17 above)
Comment (jadb) [inline tests/TestCase/Mailer/MailerTest.php]: Agreed. Will do.

File: src/Mailer/Mailer.php [hunk 19]
```diff
@@ -0,0 +1,249 @@
...
+
+    /**
+     * Checks if the property exists.
+     *
+     * @param string $offset Property name.
+     * @return bool True if it exists.
+     */
+    public function offsetExists($offset)
+    {
+        return in_array($offset, get_object_vars($this)) ||
```
Comment (markstory) [inline src/Mailer/Mailer.php]: Would `property_exists()` work here?

File: tests/test_app/TestApp/Mailer/TestMailer.php [hunk 20]
```diff
@@ -0,0 +1,32 @@
+<?php
+/**
+ * Test Suite Test App Logging stream class.
```
Comment (markstory) [inline tests/test_app/TestApp/Mailer/TestMailer.php]: I would just remove this.

**Comment by bcrowe** [comment]
Are there still plans for a MailerAwareTrait?

**Comment by jadb** [comment]
@bcrowe there is! Thanks for pointing it out, totally slipped through the cracks! Will try finishing that today.

File: src/Mailer/MailerAwareTrait.php [hunk 21]
```diff
@@ -0,0 +1,64 @@
...
+     *
+     * @var array
+     */
+    protected $_mailers = [];
+
+    /**
+     * Returns a mailer instance.
+     *
+     * @param string $name Mailer's name.
+     * @param Cake\Mailer\Email|null $email Email instance.
```
Comment (ADmad) [inline src/Mailer/MailerAwareTrait.php]: Namespaces should start with `\`

File: src/Mailer/MailerAwareTrait.php [hunk 22]
```diff
@@ -0,0 +1,64 @@
...
+            return $this->_mailers[$name];
+        }
+
+        if (is_null($email)) {
+            $email = new Email();
+        }
+
+        $className = App::className($name, 'Mailer', 'Mailer');
+
+        if (empty($className) || !class_exists($className)) {
```
Comment (ADmad) [inline src/Mailer/MailerAwareTrait.php]: `App::className()` already does a `class_exits()` check.

File: tests/TestCase/Mailer/MailerAwareTraitTest.php [hunk 23]
```diff
@@ -0,0 +1,62 @@
...
+
+}
+
+/**
+ * MailerAwareTrait test case
+ */
+class MailerAwareTraitTest extends TestCase
+{
+
+
```
Comment (ADmad) [inline tests/TestCase/Mailer/MailerAwareTraitTest.php]: Extra empty line

File: src/Mailer/MailerAwareTrait.php [hunk 24]
```diff
@@ -0,0 +1,64 @@
...
+     * @return \Cake\Mailer\Mailer
+     * @throws \Cake\Mailer\Exception\MissingMailerException if undefined mailer class.
+     */
+    public function getMailer($name, Email $email = null)
+    {
+        if (isset($this->_mailers[$name])) {
+            return $this->_mailers[$name];
+        }
+
+        if (is_null($email)) {
```
Comment (markstory) [inline src/Mailer/MailerAwareTrait.php]: `$email === null` ?

File: src/Mailer/MailerAwareTrait.php [hunk 25]
```diff
@@ -0,0 +1,64 @@
...
+     * @param string $name Mailer's name.
+     * @param \Cake\Mailer\Email|null $email Email instance.
+     * @return \Cake\Mailer\Mailer
+     * @throws \Cake\Mailer\Exception\MissingMailerException if undefined mailer class.
+     */
+    public function getMailer($name, Email $email = null)
+    {
+        if (isset($this->_mailers[$name])) {
+            return $this->_mailers[$name];
+        }
```
Comment (markstory) [inline src/Mailer/MailerAwareTrait.php]: Is there any downside to always making new mailer instances? Is sharing state something people would expect?

Comment (jadb) [inline src/Mailer/MailerAwareTrait.php]: Both are fine by me. I can think of instances where I share the same instance but mostly in the same shell/controller so there are other ways to handle that.

File: src/Mailer/Mailer.php [hunk 26]
```diff
@@ -0,0 +1,249 @@
...
+ */
+namespace Cake\Mailer;
+
+use ArrayAccess;
+use Cake\Datasource\ModelAwareTrait;
+use Cake\Event\EventListenerInterface;
+use Cake\Mailer\Exception\MissingActionException;
+use Cake\Utility\Inflector;
+
+abstract class Mailer implements ArrayAccess, EventListenerInterface
```
Comment (lorenzo) [inline src/Mailer/Mailer.php]: We usually add a header doc to each class explaining what it is

File: src/Mailer/MailerAwareTrait.php (hunk 25 above)
Comment (markstory) [inline src/Mailer/MailerAwareTrait.php]: Thanks for defaulting to not sharing state. I think that will be more transparent to users. We can easily add state sharing later. While its harder to remove.

File: src/Mailer/Mailer.php (hunk 26 above)
Comment (jadb) [inline src/Mailer/Mailer.php]: Right, am so bad with documentation. Will do it.

**Comment by markstory** [comment]
I added some doc blocks :smile:

**Comment by jadb** [comment]
@markstory thanks!

**Comment by lorenzo** [comment]
Great job @jadb :clap:

**Comment by lorenzo** [comment]
So we now need docs and the bake task :)

**Comment by jadb** [comment]
Thanks @lorenzo and to everyone else who helped reviewing/suggesting improvements.

@bcrowe started on docs. And yes, there are the bake tasks missing.
